connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -812,6 +812,7 @@ public void stop() {
         ThreadUtils.shutdownExecutorServiceQuietly(herderExecutor, herderExecutorTimeoutMs(), TimeUnit.MILLISECONDS);
         ThreadUtils.shutdownExecutorServiceQuietly(forwardRequestExecutor, FORWARD_REQUEST_SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
         ThreadUtils.shutdownExecutorServiceQuietly(startAndStopExecutor, START_AND_STOP_SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        stopServices();
```

Comment:
Would putting `stopServices` in the `finally` block in `halt` achieve the same thing? I'm a little worried about this exacerbating ungraceful shutdowns in scenarios other than the one we're trying to address with this change.

Reply:
> Would putting stopServices in the finally block in halt achieve the same thing?

Do you mean `finally` in `DistributedHerder#run`? `halt` is never called if `startServices` throws.

I considered putting `stopServices` in `catch` in `DistributedHerder#run`, but I saw that `stopServices` was already called in `StandaloneHerder#stop` and followed the same pattern.

> I'm a little worried about this exacerbating ungraceful shutdowns in scenarios other than the one we're trying to address with this change.

For ungraceful shutdowns that end with the herder thread throwing an exception, it should still call `exit` and kill the process, without ever running this code.

For ungraceful shutdowns that don't throw an exception, i suppose the herder could block and `ThreadUtils.shutdownExecutorServiceQuietly(herderExecutor, ...)` would time out, and this `stopServices` would close the KafkaBasedLogs underneath the running herder, which could be bad.

Reply:
> Do you mean `finally` in `DistributedHerder#run`? `halt` is never called if `startServices` throws.

Ugh sorry, yes. This is my reward for rushing a review before I close the laptop for the day!

> For ungraceful shutdowns that don't throw an exception, i suppose the herder could block and `ThreadUtils.shutdownExecutorServiceQuietly(herderExecutor, ...)` would time out, and this stopServices would close the KafkaBasedLogs underneath the running herder, which could be bad.

Yeah, this is what I'm afraid of.

Alternatives also include wrapping `startServices` with try/catch logic to automatically invoke `stopServices` if anything goes wrong (and then proceed to throw the original exception).

Also, looking at this once more--doesn't the `herderMetrics` field never get closed in the same scenario that we're addressing with this PR?

Reply:
BTW, this is starting to feel similar to the problem touched on in https://github.com/apache/kafka/pull/11608. Generally, both involve failure to clean up resources when operations fail in the same scope that the resources were initialized in.

Reply:
I changed this to move the stopServices from stop() to the `catch` block in run. This is because `finally` doesn't execute if exit() is called in the catch block, which is typical when in production.

I moved the stop/close for all of the non-started resources to closeResources, which is always called even if stopServices() throws. Before herderMetrics and member were only closed on the halt() happy-path, now they are called if an exception kills the herder thread too.